File: docs/test/md_testing.py
#!/usr/bin/env python3

import glob
from typing import Iterator, List
from pathlib import Path

glob_string = "../src/**/*.md"
excluded_globs = [
    "../src/fts.md",
    "../src/embedding.md",
    "../src/examples/*.md",
    "../src/integrations/*.md",
    "../src/guides/tables.md",
    "../src/python/duckdb.md",
    "../src/embeddings/*.md",
    "../src/concepts/*.md",
    "../src/ann_indexes.md",
    "../src/basic.md",
    "../src/hybrid_search/hybrid_search.md",
    "../src/reranking/*.md",
    "../src/guides/tuning_retrievers/*.md",
]

python_prefix = "py"
python_file = ".py"
python_folder = "python"

files = glob.glob(glob_string, recursive=True)
excluded_files = [
    f
    for excluded_glob in excluded_globs
    for f in glob.glob(excluded_glob, recursive=True)
]


def yield_lines(lines: Iterator[str], prefix: str, suffix: str):
    in_code_block = False
    # Python code has strict indentation
    strip_length = 0
    skip_test = False
    for line in lines:
        if "skip-test" in line:
            skip_test = True
        if line.strip().startswith(prefix + python_prefix):
            in_code_block = True
            strip_length = len(line) - len(line.lstrip())
        elif in_code_block and line.strip().startswith(suffix):
            in_code_block = False
            if not skip_test:
                yield "\n"
            skip_test = False
        elif in_code_block:
            if not skip_test:
                yield line[strip_length:]


def wrap_async(lines: List[str]) -> List[str]:
    # Indent all the lines
    lines = ["    " + line for line in lines]
    # Put all lines in `async def main():`
    lines = ["async def main():\n"] + lines
    # Put `import asyncio\n asyncio.run(main())` at the end
    lines = lines + ["\n", "import asyncio\n", "asyncio.run(main())\n"]
    return lines


for file in filter(lambda file: file not in excluded_files, files):
    with open(file, "r") as f:
        lines = list(yield_lines(iter(f), "```", "```"))

    if len(lines) > 0:
        if any("await" in line for line in lines):
            lines = wrap_async(lines)

        print(lines)
        out_path = (
            Path(python_folder)
            / Path(file).name.strip(".md")
            / (Path(file).name.strip(".md") + python_file)
        )
        print(out_path)
        out_path.parent.mkdir(exist_ok=True, parents=True)
        with open(out_path, "w") as out:
            out.writelines(lines)

File: docs/test/md_testing.js
const glob = require("glob");
const fs = require("fs");
const path = require("path");

const globString = "../src/**/*.md";

const excludedGlobs = [
  "../src/fts.md",
  "../src/embedding.md",
  "../src/examples/*.md",
  "../src/guides/tables.md",
  "../src/guides/storage.md",
  "../src/embeddings/*.md",
];

const nodePrefix = "javascript";
const nodeFile = ".js";
const nodeFolder = "node";
const asyncPrefix = "(async () => {\n";
const asyncSuffix = "})();";

function* yieldLines(lines, prefix, suffix) {
  let inCodeBlock = false;
  for (const line of lines) {
    if (line.trim().startsWith(prefix + nodePrefix)) {
      inCodeBlock = true;
    } else if (inCodeBlock && line.trim().startsWith(suffix)) {
      inCodeBlock = false;
      yield "\n";
    } else if (inCodeBlock) {
      yield line;
    }
  }
}

const files = glob.sync(globString, { recursive: true });
const excludedFiles = glob.sync(excludedGlobs, { recursive: true });

for (const file of files.filter((file) => !excludedFiles.includes(file))) {
  const lines = [];
  const data = fs.readFileSync(file, "utf-8");
  const fileLines = data.split("\n");

  for (const line of yieldLines(fileLines, "```", "```")) {
    lines.push(line);
  }

  if (lines.length > 0) {
    const fileName = path.basename(file, ".md");
    const outPath = path.join(nodeFolder, fileName, `${fileName}${nodeFile}`);
    console.log(outPath)
    fs.mkdirSync(path.dirname(outPath), { recursive: true });
    fs.writeFileSync(outPath, asyncPrefix + "\n" + lines.join("\n") + asyncSuffix);
  }
}